### ISSUE TITLE
DefaultClusterMembershipServiceTest#testClusterService is flaky

### ISSUE THREAD

**Issue by npepinpe** [camunda-cloud/zeebe#7176]
**Summary**

- How often does the test fail? twice
- Does it block your work? no
- Do we suspect that it is a real failure? doubtful

**Failures**

<details><summary>Example assertion failure</summary>
<pre>
java.lang.AssertionError: expected:<4> but was:<3>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at io.atomix.cluster.impl.DefaultClusterMembershipServiceTest.testClusterService(DefaultClusterMembershipServiceTest.java:185)
</pre>
</details>

**Logs**

<details><summary>Logs</summary>
<pre>
Standard Output
07:14:33.474 [] INFO  io.atomix.cluster.discovery.BootstrapDiscoveryProvider - Joined
07:14:33.475 [] INFO  io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Member activated: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:33.477 [] INFO  io.atomix.cluster.protocol.HeartbeatMembershipProtocol - Started
07:14:33.477 [] INFO  io.atomix.cluster.impl.DefaultClusterMembershipService - Started
07:14:33.478 [] INFO  io.atomix.cluster.discovery.BootstrapDiscoveryProvider - Joined
07:14:33.478 [] INFO  io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Member activated: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:33.478 [] INFO  io.atomix.cluster.protocol.HeartbeatMembershipProtocol - Started
07:14:33.478 [] INFO  io.atomix.cluster.impl.DefaultClusterMembershipService - Started
07:14:33.478 [] INFO  io.atomix.cluster.discovery.BootstrapDiscoveryProvider - Joined
07:14:33.479 [] INFO  io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Member activated: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:33.479 [] INFO  io.atomix.cluster.protocol.HeartbeatMembershipProtocol - Started
07:14:33.479 [] INFO  io.atomix.cluster.impl.DefaultClusterMembershipService - Started
07:14:33.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Sending heartbeat: Member{id=1, address=localhost:1, properties={}}
07:14:33.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Sending heartbeat: Member{id=1, address=localhost:1, properties={}}
07:14:33.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Sending heartbeat: Member{id=2, address=localhost:2, properties={}}
07:14:33.486 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Sending heartbeat: Member{id=2, address=localhost:2, properties={}}
07:14:33.487 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Sending heartbeat: Member{id=3, address=localhost:3, properties={}}
07:14:33.488 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Sending heartbeat: Member{id=3, address=localhost:3, properties={}}
07:14:33.518 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Received heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:33.518 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Received heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:33.518 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Received heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:33.556 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Received heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:33.556 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Received heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:33.556 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Received heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:34.477 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Sending heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:34.478 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Sending heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:34.478 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Received heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:34.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Sending heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:34.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Received heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:34.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Sending heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:34.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Received heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:34.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Sending heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:34.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Sending heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:34.481 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Received heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:34.481 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Received heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:34.481 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Received heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:35.477 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Sending heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:35.478 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Sending heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:35.478 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Received heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:35.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Received heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:35.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Sending heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:35.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Sending heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:35.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Received heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:35.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Sending heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:35.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Sending heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:35.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Received heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:35.481 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Received heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:35.481 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Received heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:36.477 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Sending heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:36.478 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Sending heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:36.478 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Received heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:36.478 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Received heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:36.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Sending heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:36.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Sending heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:36.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Received heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:36.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Sending heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:36.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Sending heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:36.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Received heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:36.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Received heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:36.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Received heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:37.477 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Sending heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:37.478 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Sending heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:37.478 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Received heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:37.478 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Received heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:37.478 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Sending heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:37.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Sending heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:37.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Received heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:37.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Received heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:37.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Sending heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:37.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Sending heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:37.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Received heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:37.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Received heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:38.477 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Sending heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:38.478 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Sending heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:38.478 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Received heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:38.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Received heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:38.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Sending heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:38.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Sending heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:38.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Sending heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:38.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Sending heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:38.481 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Received heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:38.481 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Received heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:38.482 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Received heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:38.482 [] INFO  io.atomix.cluster.discovery.BootstrapDiscoveryProvider - Joined
07:14:38.482 [] INFO  io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 4 - Member activated: Member{id=4, address=localhost:4, host=localhost, properties={}}
07:14:38.482 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Received heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:38.483 [] INFO  io.atomix.cluster.protocol.HeartbeatMembershipProtocol - Started
07:14:38.483 [] INFO  io.atomix.cluster.impl.DefaultClusterMembershipService - Started
07:14:38.483 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 4 - Sending heartbeat: Member{id=1, address=localhost:1, properties={}}
07:14:38.484 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 4 - Sending heartbeat: Member{id=2, address=localhost:2, properties={}}
07:14:38.484 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 4 - Sending heartbeat: Member{id=3, address=localhost:3, properties={}}
07:14:38.484 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Received heartbeat: Member{id=4, address=localhost:4, host=localhost, properties={}}
07:14:38.484 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Received heartbeat: Member{id=4, address=localhost:4, host=localhost, properties={}}
07:14:38.484 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Received heartbeat: Member{id=4, address=localhost:4, host=localhost, properties={}}
07:14:39.477 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Sending heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:39.478 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Sending heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:39.478 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Sending heartbeat: Member{id=4, address=localhost:4, host=localhost, properties={}}
07:14:39.478 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Received heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:39.478 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Received heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:39.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 4 - Received heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:39.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Sending heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:39.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Sending heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:39.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Sending heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:39.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Sending heartbeat: Member{id=4, address=localhost:4, host=localhost, properties={}}
07:14:39.479 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Sending heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:39.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Sending heartbeat: Member{id=4, address=localhost:4, host=localhost, properties={}}
07:14:39.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 4 - Received heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:39.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Received heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:39.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Received heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:39.480 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Received heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:39.481 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Received heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:39.483 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 4 - Received heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:39.485 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 4 - Sending heartbeat: Member{id=1, address=localhost:1, host=localhost, properties={}}
07:14:39.485 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 4 - Sending heartbeat: Member{id=2, address=localhost:2, host=localhost, properties={}}
07:14:39.485 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 4 - Sending heartbeat: Member{id=3, address=localhost:3, host=localhost, properties={}}
07:14:39.485 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 1 - Received heartbeat: Member{id=4, address=localhost:4, host=localhost, properties={}}
07:14:39.485 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 2 - Received heartbeat: Member{id=4, address=localhost:4, host=localhost, properties={}}
07:14:39.487 [] TRACE io.atomix.cluster.protocol.HeartbeatMembershipProtocol - 3 - Received heartbeat: Member{id=4, address=localhost:4, host=localhost, properties={}}
</pre>
</details>

**Comment by npepinpe** [comment]
I looked at the test, and here is what it currently does:

```java
  void testClusterService() {
    // builds cluster service (svc1) for member 1
    // builds cluster service (svc2) for member 2
    // builds cluster service (svc3) for member 3
    // asserts that svc1 doesn't know about anybody, even itself, via getMember
    // starts all cluster services in parallel
    // sleeps for 5 seconds
    // asserts that all services are aware of 3 members via getMembers().size()
    // asserts that all members - including localMember - of svc1 are active
    // asserts that all members of svc1 have the expected version
    // builds a new ephemeral service (svc4) for member 4
    // asserts - via svc4 - that member 4 is inactive and does not know about any members
    // starts svc4
    // sleeps for 1 second
    // asserts that all services now detect there are 4 members in the cluster
    // again asserts the versions of all members
    // stops svc1
    // sleeps for 5 seconds
    // asserts that svc2 only reports 3 members
    // asserts that svc2 does not know about member 1 anymore
    // asserts via svc2 that members 2-4 are active
    // stops svc4
    // sleeps 5 seconds
    // asserts that svc2 does not know about member 1 anymore
    // asserts via svc2 that members 2 and 3 are active
    // asserts that svc2 does not know about member 4 anymore
    // sleeps 2.5 seconds
    // asserts again that svc2 only knows 2 members
    // asserts that svc2 does not know about member 1 anymore
    // asserts via svc2 that members 2 and 3 are active
    // asserts that svc2 does not know about member 4 anymore
    // adds an event listener to svc2
    // updates the properties of member 3
    // requests the next event via the listener
    // asserts that the event is of type METADATA_CHANGED
    // asserts that the event reports the updated property
    // stops all services in parallel
  }
```

**Comment by npepinpe** [comment]
Note that its setup is to use mostly test stubs, except use the `HeartbeatMembershipProtocol` for group membership - so this is effectively mostly testing this protocol, which we don't use in production (we only use SWIM).

I would propose instead to split this into multiple tests, and actually only tests the cluster membership service. What are its responsibilities?

- Glue together the discovery protocol and the group membership protocol
- Report members accurately from these
- Keep track of its local member's reachability status
- Manage the lifecycle of the discovery service
- Listen to and forward group membership events

I would remove the use of the heartbeat membership protocol and try to remove any asynchronous I/O at all, and just use the event listeners and the lifecycle futures to synchronize things.
